src/exo/LoopIR_scheduling.py

```diff
@@ -301,42 +313,74 @@ def DoReorderStmt(f_cursor, s_cursor):
     return ir, fwd
 
 
-def DoPartitionLoop(stmt, partition_by):
-    s = stmt._node
+def DoCutLoop(loop_c, cut_point):
+    s = loop_c._node
 
     assert isinstance(s, LoopIR.Seq)
 
-    part_by = LoopIR.Const(partition_by, T.int, s.srcinfo)
-    new_hi = LoopIR.BinOp("-", s.hi, part_by, T.int, s.srcinfo)
     try:
-        Check_IsPositiveExpr(
-            stmt.get_root(),
+        Check_IsNonNegativeExpr(
+            loop_c.get_root(),
             [s],
-            LoopIR.BinOp(
-                "+", new_hi, LoopIR.Const(1, T.int, s.srcinfo), T.int, s.srcinfo
-            ),
+            LoopIR.BinOp("-", cut_point, s.lo, T.index, s.srcinfo),
         )
     except SchedulingError:
-        raise SchedulingError(
-            f"expected the new loop bound {new_hi} to be always non-negative"
+        raise SchedulingError(f"Expected `lo` <= `cut_point`")
+
+    try:
+        Check_IsNonNegativeExpr(
+            loop_c.get_root(),
+            [s],
+            LoopIR.BinOp("-", s.hi, cut_point, T.index, s.srcinfo),
         )
+    except SchedulingError:
+        raise SchedulingError(f"Expected `cut_point` <= `hi`")
+
+    ir, fwd1 = loop_c._child_node("hi")._replace(cut_point)
+    loop2 = Alpha_Rename([s.update(lo=cut_point)]).result()[0]
+    ir, fwd2 = fwd1(loop_c).after()._insert([loop2])
+    fwd = _compose(fwd2, fwd1)
+
+    return ir, fwd
+
 
-    loop1 = Alpha_Rename([s.update(hi=part_by)]).result()[0]
+def DoShiftLoop(loop_c, new_lo):
+    s = loop_c._node
+
+    assert isinstance(s, LoopIR.Seq)
+
+    try:
+        Check_IsNonNegativeExpr(
+            loop_c.get_root(),
+            [s],
+            new_lo,
+        )
+    except SchedulingError:
+        raise SchedulingError(f"Expected 0 <= `new_lo`")
```

Comment:
Is there a reason we don't want negative loops?

Reply:
My understanding is that so far we have been maintaining the invariant that
0 <= lo <= hi

I am not sure if changing it would break anything somewhere. It might be worth it to revise LoopIR to have actual classes with initializer to document those invariants in code as asserts.

Reply:
Hmm, that might be an artifact of having lo = 0 in our previous representation of for loops. In any case, it's probably not essential to support negative values right now though.